// webstomp/dist/app/app.js
(function() {
  var App, Router, Server, Signal, assign;

  Signal = require("../signal");

  Router = require("../router");

  Server = require("../server");

  assign = require("lodash").assign;

  App = (function() {
    App.events = ["error"];

    function App() {
      this.router = new Router();
      this.signals = {};
      this.constructor.events.forEach((function(_this) {
        return function(name) {
          return _this.signals[name] = new Signal();
        };
      })(this));
    }

    App.prototype.on = function(name, fn) {
      return this.addListener(name, fn);
    };

    App.prototype.addListener = function(name, fn) {
      return this.signals[name].addListener(fn);
    };

    App.prototype.removeListener = function(name, fn) {
      return this.signals[name].removeListener(fn);
    };

    App.prototype.accept = function(transport) {
      var dispatch, state;
      transport.on("error", (function(_this) {
        return function(err) {
          return _this.signals.error.emit(err);
        };
      })(this));
      state = {};
      dispatch = (function(_this) {
        return function(frame) {
          var context;
          context = Object.create(transport);
          assign(context, frame, {
            state: state
          });
          return _this.router.dispatch(context, function(err) {
            if (err) {
              return _this.signals.error.emit(err);
            }
          });
        };
      })(this);
      transport.on("frame", function(frame) {
        return dispatch(frame);
      });
      transport.on("close", function() {
        return dispatch({
          command: "DISCONNECT",
          headers: {}
        });
      });
      return null;
    };

    App.prototype.mount = function(params) {
      var server;
      server = new Server(params);
      server.on("connection", this.accept.bind(this));
      return server;
    };

    App.prototype.listen = function(port) {
      return this.mount({
        port: port
      });
    };

    return App;

  })();

  Router.commands.concat(["use"]).forEach(function(name) {
    return App.prototype[name] = function() {
      return this.router[name].apply(this.router, arguments);
    };
  });

  module.exports = App;

}).call(this);
