// webstomp/dist/frame/frame.js
(function() {
  var assign, each, fromString, ref, toString;

  ref = require("lodash"), assign = ref.assign, each = ref.each;

  toString = function(arg) {
    var body, command, headers, s;
    command = arg.command, headers = arg.headers, body = arg.body;
    headers = assign({}, headers);
    if (body) {
      headers['content-length'] = body.length;
    }
    s = command + "\n";
    each(headers, function(value, key) {
      return s += key + ":" + value + "\n";
    });
    s += "\n";
    if (body) {
      s += body;
    }
    s += "\0";
    return s;
  };

  fromString = function(s) {
    var body, command, end, frame, headers, key, line, sep, start, value;
    if (!(s && s.length > 0)) {
      return null;
    }
    if (s === "\n") {
      return null;
    }
    start = 0;
    command = null;
    while (true) {
      end = s.indexOf("\n", start);
      if (end === -1) {
        throw new Error("invalid command");
      }
      line = s.slice(start, end);
      start = end + 1;
      if (start >= s.length) {
        throw new Error("invalid command");
      }
      if (line !== "") {
        command = line;
        break;
      }
    }
    headers = {};
    while (true) {
      end = s.indexOf("\n", start);
      if (end === -1) {
        break;
      }
      line = s.slice(start, end);
      start = end + 1;
      if (line === "") {
        break;
      }
      sep = line.indexOf(":", 0);
      if (sep === -1) {
        throw new Error("invalid header: " + line);
      }
      key = line.slice(0, sep);
      if (!headers[key]) {
        value = line.slice(sep + 1);
        headers[key] = value;
      }
    }
    if (headers['content-length']) {
      headers['content-length'] = parseInt(headers['content-length'], 10);
    }
    body = (function() {
      if (headers['content-length']) {
        end = start + headers['content-length'];
        if (end > s.length) {
          throw new Error("invalid content length");
        }
        return s.slice(start, end);
      } else if (s[start] !== "\0") {
        end = s.length - 1;
        while (s[end] === "\n" || s[end] === "\0") {
          end -= 1;
          if (end <= start) {
            throw new Error("no frame terminator");
          }
        }
        return s.slice(start, end + 1);
      } else {
        return null;
      }
    })();
    frame = {
      command: command,
      headers: headers,
      body: body
    };
    return frame;
  };

  module.exports = {
    toString: toString,
    fromString: fromString
  };

}).call(this);
