// webstomp/dist/transport/transport.js
(function() {
  var Frame, Signal, Transport, fromCallback;

  Signal = require("../signal");

  Frame = require("../frame");

  fromCallback = require("bluebird").fromCallback;

  Transport = (function() {
    Transport.events = ["open", "message", "frame", "heartbeat", "error", "close"];

    function Transport(ws) {
      this.ws = ws;
      this.heartbeat = null;
      this.signals = {};
      this.constructor.events.forEach((function(_this) {
        return function(name) {
          return _this.signals[name] = new Signal();
        };
      })(this));
      this.ws.on("open", (function(_this) {
        return function() {
          return _this.signals.open.emit(true);
        };
      })(this));
      this.ws.on("message", (function(_this) {
        return function(message) {
          return _this.receiveMessage(message);
        };
      })(this));
      this.ws.on("error", (function(_this) {
        return function(err) {
          return _this.signals.error.emit(err);
        };
      })(this));
      this.ws.on("close", (function(_this) {
        return function() {
          clearInterval(_this.heartbeat);
          return _this.signals.close.emit(true);
        };
      })(this));
    }

    Transport.prototype.on = function(name, fn) {
      return this.addListener(name, fn);
    };

    Transport.prototype.addListener = function(name, fn) {
      return this.signals[name].addListener(fn);
    };

    Transport.prototype.removeListener = function(name, fn) {
      return this.signals[name].removeListener(fn);
    };

    Transport.prototype.receiveMessage = function(message) {
      var err, error, frame;
      if (/^\s*$/.test(message)) {
        return this.signals.heartbeat.emit(true);
      } else {
        try {
          frame = Frame.fromString(message);
          return this.signals.frame.emit(frame);
        } catch (error) {
          err = error;
          this.signals.error.emit(err);
          return this.close();
        }
      }
    };

    Transport.prototype.sendFrame = function(frame) {
      var message;
      message = Frame.toString(frame);
      return fromCallback((function(_this) {
        return function(onError) {
          return _this.ws.send(message, onError);
        };
      })(this));
    };

    Transport.prototype.sendHeartbeat = function() {
      return fromCallback((function(_this) {
        return function(onError) {
          return _this.ws.send("\n", onError);
        };
      })(this));
    };

    Transport.prototype.autoHeartbeat = function(ms) {
      var sendHeartbeat;
      sendHeartbeat = this.sendHeartbeat.bind(this);
      return this.heartbeat = setInterval(sendHeartbeat, ms);
    };

    Transport.prototype.close = function() {
      return this.ws.close();
    };

    return Transport;

  })();

  module.exports = Transport;

}).call(this);
