// webstomp/dist/server-transport/server-transport.js
(function() {
  var ServerTransport, Transport, assertHeaders,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Transport = require("../transport");

  ServerTransport = (function(superClass) {
    extend(ServerTransport, superClass);

    function ServerTransport() {
      return ServerTransport.__super__.constructor.apply(this, arguments);
    }

    ServerTransport.prototype.connected = function(headers) {
      assertHeaders(headers, ["version"]);
      return this.sendFrame({
        command: "CONNECTED",
        headers: headers
      });
    };

    ServerTransport.prototype.message = function(headers, body) {
      assertHeaders(headers, ["destination", "subscription", "message-id"]);
      return this.sendFrame({
        command: "MESSAGE",
        headers: headers,
        body: body
      });
    };

    ServerTransport.prototype.receipt = function(headers) {
      assertHeaders(headers, ["receipt-id"]);
      return this.sendFrame({
        command: "RECEIPT",
        headers: headers
      });
    };

    ServerTransport.prototype.error = function(headers, body) {
      assertHeaders(headers, ["message"]);
      return this.sendFrame({
        command: "ERROR",
        headers: headers,
        body: body
      });
    };

    return ServerTransport;

  })(Transport);

  assertHeaders = function(headers, names) {
    var i, len, name, results;
    if (!headers) {
      throw new Error("headers required");
    }
    results = [];
    for (i = 0, len = names.length; i < len; i++) {
      name = names[i];
      if (!headers[name]) {
        throw new Error(name + " header required");
      } else {
        results.push(void 0);
      }
    }
    return results;
  };

  module.exports = ServerTransport;

}).call(this);
