// webstomp/dist/client-transport/client-transport.js
(function() {
  var ClientTransport, Transport, assertHeaders,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  Transport = require("../transport");

  ClientTransport = (function(superClass) {
    extend(ClientTransport, superClass);

    function ClientTransport() {
      return ClientTransport.__super__.constructor.apply(this, arguments);
    }

    ClientTransport.prototype.connect = function(headers) {
      assertHeaders(headers, ["accept-version", "host"]);
      return this.sendFrame({
        command: "CONNECT",
        headers: headers
      });
    };

    ClientTransport.prototype.stomp = function(headers) {
      return this.connect(headers);
    };

    ClientTransport.prototype.send = function(headers, body) {
      assertHeaders(headers, ["destination"]);
      return this.sendFrame({
        command: "SEND",
        headers: headers,
        body: body
      });
    };

    ClientTransport.prototype.subscribe = function(headers) {
      assertHeaders(headers, ["destination", "id"]);
      return this.sendFrame({
        command: "SUBSCRIBE",
        headers: headers
      });
    };

    ClientTransport.prototype.unsubscribe = function(headers) {
      assertHeaders(headers, ["id"]);
      return this.sendFrame({
        command: "UNSUBSCRIBE",
        headers: headers
      });
    };

    ClientTransport.prototype.disconnect = function(headers) {
      return this.sendFrame({
        command: "DISCONNECT",
        headers: headers
      });
    };

    return ClientTransport;

  })(Transport);

  assertHeaders = function(headers, names) {
    var i, len, name, results;
    if (!headers) {
      throw new Error("headers required");
    }
    results = [];
    for (i = 0, len = names.length; i < len; i++) {
      name = names[i];
      if (!headers[name]) {
        throw new Error(name + " header required");
      } else {
        results.push(void 0);
      }
    }
    return results;
  };

  module.exports = ClientTransport;

}).call(this);
