// webstomp/dist/server/server.js
(function() {
  var Server, ServerTransport, Signal, WebSocket;

  WebSocket = require("ws");

  Signal = require("../signal");

  ServerTransport = require("../server-transport");

  Server = (function() {
    Server.events = ["listening", "connection", "error"];

    function Server(params) {
      this.signals = {};
      this.constructor.events.forEach((function(_this) {
        return function(name) {
          return _this.signals[name] = new Signal();
        };
      })(this));
      this.server = WebSocket.createServer(params, (function(_this) {
        return function(ws) {
          return _this.receiveConnection(ws);
        };
      })(this));
      this.server.on("listening", (function(_this) {
        return function() {
          return _this.signals.listening.emit(true);
        };
      })(this));
      this.server.on("error", (function(_this) {
        return function(err) {
          return _this.signals.error.emit(err);
        };
      })(this));
    }

    Server.prototype.on = function(name, fn) {
      return this.addListener(name, fn);
    };

    Server.prototype.addListener = function(name, fn) {
      return this.signals[name].addListener(fn);
    };

    Server.prototype.removeListener = function(name, fn) {
      return this.signals[name].removeListener(fn);
    };

    Server.prototype.receiveConnection = function(ws) {
      var transport;
      transport = new ServerTransport(ws);
      return this.signals.connection.emit(transport);
    };

    Server.prototype.close = function() {
      return this.server.close();
    };

    return Server;

  })();

  module.exports = Server;

}).call(this);
